File: trifecta-ios/Trifecta/Core/Networking/Auth/PairingFlow.swift
import Foundation

struct PairingResult: Sendable {
    let bearerToken: String
}

enum PairingFlow {
    struct WebSocketToken: Sendable {
        let token: String
    }

    static func fetchEnvironment(serverURL: URL) async throws -> EnvironmentDescriptor {
        let url = serverBaseURL(from: serverURL).appendingPathComponent(".well-known/t3/environment")
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.timeoutInterval = 10
        let (data, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse else {
            throw T3Error.pairingFailed("No response from server")
        }
        guard http.statusCode == 200 else {
            throw T3Error.pairingFailed(
                "Environment request failed with status \(http.statusCode): \(errorMessage(from: data))"
            )
        }
        guard let descriptor = EnvironmentDescriptor.decodeLenient(data) else {
            throw T3Error.pairingFailed("Invalid environment descriptor")
        }
        return descriptor
    }

    static func exchangeToken(serverURL: URL, oneTimeToken: String) async throws -> PairingResult {
        let url = serverBaseURL(from: serverURL).appendingPathComponent("api/auth/bootstrap/bearer")
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.timeoutInterval = 15
        let body: [String: Any] = ["credential": oneTimeToken]
        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        let (data, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse else {
            throw T3Error.pairingFailed("No response from server")
        }
        if http.statusCode == 401 || http.statusCode == 403 {
            throw T3Error.pairingFailed("Pairing token rejected: \(errorMessage(from: data))")
        }
        guard (200...299).contains(http.statusCode) else {
            throw T3Error.pairingFailed(
                "Pairing failed with status \(http.statusCode): \(errorMessage(from: data))"
            )
        }

        if let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] {
            if let token = (json["sessionToken"] ?? json["token"] ?? json["bearer"]) as? String,
               !token.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
                return PairingResult(bearerToken: token)
            }
        }

        throw T3Error.pairingFailed("Pairing response did not include a bearer session token")
    }

    static func issueWebSocketToken(serverURL: URL, bearerToken: String) async throws -> WebSocketToken {
        let url = serverBaseURL(from: serverURL).appendingPathComponent("api/auth/ws-token")
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("Bearer \(bearerToken)", forHTTPHeaderField: "Authorization")
        request.timeoutInterval = 10

        let (data, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse else {
            throw T3Error.pairingFailed("No response from server")
        }
        guard (200...299).contains(http.statusCode) else {
            throw T3Error.pairingFailed(
                "WebSocket token request failed with status \(http.statusCode): \(errorMessage(from: data))"
            )
        }
        guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let token = json["token"] as? String,
              !token.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            throw T3Error.pairingFailed("WebSocket token response was invalid")
        }
        return WebSocketToken(token: token)
    }

    static func parsePairingURL(_ raw: String) -> (serverURL: URL, token: String)? {
        let trimmed = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty, let comps = URLComponents(string: trimmed) else { return nil }

        if let host = comps.queryItems?.first(where: { $0.name == "host" })?.value,
           let backend = URL(string: host),
           let token = parseToken(from: comps) {
            return (serverBaseURL(from: backend), token)
        }

        if let token = parseToken(from: comps),
           let scheme = comps.scheme, let host = comps.host {
            var direct = URLComponents()
            direct.scheme = scheme
            direct.host = host
            direct.port = comps.port
            return direct.url.map { ($0, token) }
        }
        return nil
    }

    static func serverBaseURL(from url: URL) -> URL {
        guard var comps = URLComponents(url: url, resolvingAgainstBaseURL: false) else {
            return url
        }
        comps.path = ""
        comps.query = nil
        comps.fragment = nil
        return comps.url ?? url
    }

    private static func parseToken(from comps: URLComponents) -> String? {
        if let token = parseTokenFromFragment(comps.fragment) {
            return token
        }
        return comps.queryItems?.first(where: { $0.name == "token" })?.value?
            .trimmingCharacters(in: .whitespacesAndNewlines)
    }

    private static func parseTokenFromFragment(_ fragment: String?) -> String? {
        guard let fragment,
              let comps = URLComponents(string: "?\(fragment)"),
              let token = comps.queryItems?.first(where: { $0.name == "token" })?.value?
                  .trimmingCharacters(in: .whitespacesAndNewlines),
              !token.isEmpty else { return nil }
        return token
    }

    private static func errorMessage(from data: Data) -> String {
        if let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
           let message = json["error"] as? String,
           !message.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            return message
        }
        if let text = String(data: data, encoding: .utf8)?
            .trimmingCharacters(in: .whitespacesAndNewlines),
           !text.isEmpty {
            return text
        }
        return "empty response"
    }
}

File: trifecta-android/app/src/main/java/com/belweave/trifecta/core/auth/PairingFlow.kt
package com.belweave.trifecta.core.auth

import com.belweave.trifecta.core.models.EnvironmentDescriptor
import com.belweave.trifecta.core.models.T3Json
import com.belweave.trifecta.core.models.asObjectOrNull
import com.belweave.trifecta.core.models.str
import com.belweave.trifecta.core.networking.T3Error
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import okhttp3.Call
import okhttp3.Callback
import okhttp3.HttpUrl
import okhttp3.HttpUrl.Companion.toHttpUrlOrNull
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.Response
import java.io.IOException
import java.net.URI
import java.util.concurrent.TimeUnit
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException

/**
 * Performs the bearer-token pairing dance against a T3 server.
 * Mirrors `PairingFlow.swift` 1:1 in shape and error semantics.
 */
class PairingFlow(private val httpClient: OkHttpClient = defaultClient()) {

    data class WebSocketToken(val token: String)
    data class PairingResult(val bearerToken: String)

    suspend fun fetchEnvironment(serverURL: URI): EnvironmentDescriptor =
        withContext(Dispatchers.IO) {
            val url = serverBaseURL(serverURL).toHttpUrlOrNull()
                ?.newBuilder()
                ?.addPathSegments(".well-known/belweave/environment")
                ?.build()
                ?: throw T3Error.PairingFailed("Invalid server URL")

            val request = Request.Builder().url(url).get().build()
            httpClient.executeAsync(request).use { response ->
                if (!response.isSuccessful) {
                    throw T3Error.PairingFailed(
                        "Environment request failed with status ${response.code}: ${errorBody(response)}"
                    )
                }
                val body = response.body?.string().orEmpty()
                EnvironmentDescriptor.decodeLenient(body)
                    ?: throw T3Error.PairingFailed("Invalid environment descriptor")
            }
        }

    suspend fun exchangeToken(serverURL: URI, oneTimeToken: String): PairingResult =
        withContext(Dispatchers.IO) {
            val url = serverBaseURL(serverURL).toHttpUrlOrNull()
                ?.newBuilder()
                ?.addPathSegments("api/auth/bootstrap/bearer")
                ?.build()
                ?: throw T3Error.PairingFailed("Invalid server URL")

            val payload = buildJsonObject { put("credential", oneTimeToken) }
            val body = T3Json.encodeToString(JsonObject.serializer(), payload)
                .toRequestBody(JSON_MEDIA)
            val request = Request.Builder().url(url).post(body).build()

            httpClient.executeAsync(request).use { response ->
                if (response.code == 401 || response.code == 403) {
                    throw T3Error.PairingFailed("Pairing token rejected: ${errorBody(response)}")
                }
                if (!response.isSuccessful) {
                    throw T3Error.PairingFailed(
                        "Pairing failed with status ${response.code}: ${errorBody(response)}"
                    )
                }
                val text = response.body?.string().orEmpty()
                val obj = T3Json.parseToJsonElement(text).asObjectOrNull()
                val token = obj?.str("sessionToken")
                    ?: obj?.str("token")
                    ?: obj?.str("bearer")
                if (token.isNullOrBlank()) {
                    throw T3Error.PairingFailed("Pairing response did not include a bearer session token")
                }
                PairingResult(token.trim())
            }
        }

    suspend fun issueWebSocketToken(serverURL: URI, bearerToken: String): WebSocketToken =
        withContext(Dispatchers.IO) {
            val url = serverBaseURL(serverURL).toHttpUrlOrNull()
                ?.newBuilder()
                ?.addPathSegments("api/auth/ws-token")
                ?.build()
                ?: throw T3Error.PairingFailed("Invalid server URL")

            val request = Request.Builder()
                .url(url)
                .post(EMPTY_BODY)
                .header("Authorization", "Bearer $bearerToken")
                .build()

            httpClient.executeAsync(request).use { response ->
                if (!response.isSuccessful) {
                    throw T3Error.PairingFailed(
                        "WebSocket token request failed with status ${response.code}: ${errorBody(response)}"
                    )
                }
                val text = response.body?.string().orEmpty()
                val token = T3Json.parseToJsonElement(text).asObjectOrNull()?.str("token")
                if (token.isNullOrBlank()) {
                    throw T3Error.PairingFailed("WebSocket token response was invalid")
                }
                WebSocketToken(token.trim())
            }
        }

    private fun errorBody(response: Response): String {
        val text = response.body?.string()?.trim().orEmpty()
        if (text.isEmpty()) return "empty response"
        val obj = runCatching { T3Json.parseToJsonElement(text).asObjectOrNull() }.getOrNull()
        return obj?.str("error")?.takeIf { it.isNotBlank() } ?: text
    }

    companion object {
        private val JSON_MEDIA = "application/json".toMediaType()
        private val EMPTY_BODY = ByteArray(0).toRequestBody(JSON_MEDIA)

        fun defaultClient(): OkHttpClient = OkHttpClient.Builder()
            .connectTimeout(15, TimeUnit.SECONDS)
            .readTimeout(15, TimeUnit.SECONDS)
            .writeTimeout(15, TimeUnit.SECONDS)
            .build()

        fun serverBaseURL(url: URI): URI {
            val scheme = url.scheme ?: return url
            val host = url.host ?: return url
            val port = if (url.port == -1) -1 else url.port
            val rawPath = url.rawPath.orEmpty().trimEnd('/')
            val basePath = when {
                rawPath == "/pair" -> null
                rawPath.endsWith("/pair") -> rawPath.removeSuffix("/pair").takeIf { it.isNotEmpty() }
                rawPath.isNotEmpty() -> rawPath
                else -> null
            }
            return URI(scheme, null, host, port, basePath, null, null)
        }

        /**
         * Parses pairing URLs of the form:
         *   trifecta://pair?host=https://server&token=...
         *   https://server/pair#token=...
         *   trifecta://server?token=...
         */
        fun parsePairingURL(raw: String): Pair<URI, String>? {
            val trimmed = raw.trim()
            if (trimmed.isEmpty()) return null
            val uri = runCatching { URI(trimmed) }.getOrNull() ?: return null

            val queryParams = parseQuery(uri.rawQuery)
            val fragmentParams = parseQuery(uri.rawFragment)

            queryParams["host"]?.let { hostRaw ->
                val backend = runCatching { URI(hostRaw) }.getOrNull()
                val token = parseToken(queryParams, fragmentParams)
                if (backend != null && token != null) {
                    return serverBaseURL(backend) to token
                }
            }

            val token = parseToken(queryParams, fragmentParams)
            val scheme = uri.scheme
            val host = uri.host
            if (token != null && scheme != null && host != null) {
                return serverBaseURL(uri) to token
            }
            return null
        }

        private fun parseToken(query: Map<String, String>, fragment: Map<String, String>): String? {
            val fromFragment = fragment["token"]?.trim()?.takeIf { it.isNotEmpty() }
            if (fromFragment != null) return fromFragment
            return query["token"]?.trim()?.takeIf { it.isNotEmpty() }
        }

        private fun parseQuery(raw: String?): Map<String, String> {
            if (raw.isNullOrEmpty()) return emptyMap()
            return raw.split('&').mapNotNull { pair ->
                val idx = pair.indexOf('=')
                if (idx < 0) return@mapNotNull null
                val key = decode(pair.substring(0, idx))
                val value = decode(pair.substring(idx + 1))
                key to value
            }.toMap()
        }

        private fun decode(text: String): String =
            runCatching { java.net.URLDecoder.decode(text, "UTF-8") }.getOrDefault(text)
    }
}

/** Bridge OkHttp's blocking `execute` into a coroutine-friendly cancellable suspend call. */
private suspend fun OkHttpClient.executeAsync(request: Request): Response =
    suspendCancellableCoroutine { cont ->
        val call = newCall(request)
        cont.invokeOnCancellation { runCatching { call.cancel() } }
        call.enqueue(object : Callback {
            override fun onFailure(call: Call, e: IOException) {
                if (cont.isActive) cont.resumeWithException(e)
            }
            override fun onResponse(call: Call, response: Response) {
                if (cont.isActive) cont.resume(response)
            }
        })
    }
